### ISSUE TITLE
False positive for unnecessary_lambdas

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#473]
False positive for https://github.com/flutter/flutter/blob/aac8762a7a16d931eebde40e7179f9913867ca8b/packages/flutter_test/lib/src/controller.dart#L172

```dart
return finder.evaluate().map((Element element) => _stateOf<T>(element, finder));
```

@JPaulsen

**Comment by JPaulsen** [comment]
@a14n are you sure? I could not reproduce the bug, how do you run the linter in flutter?

**Comment by a14n** [comment]
Once you have the flutter repo, you need to change `packages/flutter_tools/pubspec.yaml` to override linter dependency (eithier with a git reference or with a path reference):

```
diff --git a/packages/flutter_tools/pubspec.yaml b/packages/flutter_tools/pubspec.yaml
index feeaac5..5c60da8 100644
--- a/packages/flutter_tools/pubspec.yaml
+++ b/packages/flutter_tools/pubspec.yaml
@@ -52,3 +52,5 @@ dependency_overrides:
     path: ../../bin/cache/dart-sdk/lib/analyzer
   front_end:
     path: ../../bin/cache/dart-sdk/lib/front_end
+  linter:
+    path: ../../../../dart-lang/linter
```

You may have to remove the flutter_tools snapshot to take this change into account with `rm bin/cache/flutter_tools.snapshot`.

Now you can add any linter rule to `.analysis_options_repo` and run `flutter analyze --flutter-repo` to display lints.

I already send a [PR](https://github.com/flutter/flutter/pull/8595/files) to fix some lint but there are actually other false positives. See the Hixie's comments.

**Comment by a14n** [comment]
Basically the lint should apply only when the function content contains only one method call with the same parameter count:

```dart
call((){ a.m1(); }); // LINT
call((p){ a.m2(p); }) // LINT
call((){ a?.m1(); }) // when a is null, not the same as call(a?.m1)
call((){ a.b.m1(); }) // when b is mutable, not the same as call(a.b.m1)
```

**Comment by alexeieleusis** [comment]
https://github.com/flutter/flutter/pull/8595/files#diff-a58f6bf53c00b37f422f3cccc4e906a8R35

**Comment by a14n** [comment]
Actually here's another false positive:

```dart
main() {
  Function makeCallable(void f()) => f;
  
  var f = () => print('a');
  
  // r1 is linted (r2 is the result to remove the lint)
  var r1 = makeCallable((){
    f();
  });
  var r2 = makeCallable(f);
  
  f = () => print('b');
  
  r1(); // prints b
  r2(); // prints a
}
```

**Comment by a14n** [comment]
IMHO the lint should apply only to final var/field with only one method call or to function/method (`call((){ this.m1(); })`, `call((){ f1(); })`...)

**Comment by JPaulsen** [comment]
@a14n , @bwilkerson, with @alexeieleusis we think that that is not enough to conclude that the method will be immutable, in fact, an object is immutable if and only if all theirs camp are immutable (recursively).

For example, I can have: 
```
final List a = [];
```
In this case a is final but not immutable.

**Comment by a14n** [comment]
I don't see how your example with `a` can lead to false positives with [the rule](https://github.com/dart-lang/linter/issues/473#issuecomment-284643975):

>the lint should apply only to final var/field with only one method call or to function/method

Can you give me an example where this rule fails ?
